StripeConnect/StripeConnect/Source/Internal/AuthenticatedWebView/AuthenticatedWebViewManager.swift

```diff
@@ -0,0 +1,79 @@
+//
+//  AuthenticatedWebViewManager.swift
+//  StripeConnect
+//
+//  Created by Mel Ludowise on 10/15/24.
+//
+
+import AuthenticationServices
+
+/// Singleton helper to manage
+class AuthenticatedWebViewManager: NSObject {
+    typealias SessionFactory = (
+        _ url: URL,
+        _ callbackURLScheme: String?,
+        _ completionHandler: @escaping ASWebAuthenticationSession.CompletionHandler
+    ) -> ASWebAuthenticationSession
+
+    /// Used to dependency inject `ASWebAuthenticationSession.init` in tests
+    private let sessionFactory: SessionFactory
+
+    /// Window to present the session in
+    private weak var window: UIWindow?
+
+    /// The currently presented auth session, if there is one
+    weak var authSession: ASWebAuthenticationSession?
+
+    init(sessionFactory: @escaping SessionFactory = ASWebAuthenticationSession.init) {
+        self.sessionFactory = sessionFactory
+    }
+
+    /// Returns the redirect URL or nil if the user cancelled the flow
+    func present(with url: URL, in window: UIWindow?) async throws -> URL? {
+        guard authSession == nil else {
+            throw AuthenticatedWebViewError.alreadyPresenting
+        }
+        guard let window else {
+            throw AuthenticatedWebViewError.noWindow
+        }
+        self.window = window
+
+        let returnUrl: URL? = try await withCheckedThrowingContinuation { continuation in
+            let authSession = sessionFactory(url, StripeConnectConstants.authenticatedWebViewReturnUrlScheme) { returnUrl, error in
+
+                if let authenticationSessionError = error as? ASWebAuthenticationSessionError,
+                   authenticationSessionError.code == .canceledLogin {
+                    // The user either selected "Cancel" in the initial modal
+                    // prompting them to "Sign In" or they hit the "Cancel"
+                    // button in presented browser view
+                    continuation.resume(returning: nil)
+                    return
+                } else if let error {
+                    continuation.resume(throwing: error)
+                    return
+                }
+
+                continuation.resume(returning: returnUrl)
+            }
+            authSession.presentationContextProvider = self
+
+            self.authSession = authSession
+
+            guard authSession.canStart,
+                  authSession.start() else {
+                continuation.resume(throwing: AuthenticatedWebViewError.cannotStartSession)
+                return
+            }
+        }
+
+        return returnUrl
+    }
+}
+
+// MARK: - ASWebAuthenticationPresentationContextProviding
+
+extension AuthenticatedWebViewManager: ASWebAuthenticationPresentationContextProviding {
+    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
+        return window ?? .init()
```

Comment:
It's possible that calling `UIWindow.init()` could cause cause a fatalError since this is called off the main thread, but we do check that `window` is non-nil before starting the session, so I don't think this could realistically happen. 

When I tested locally and changed this to always return `.init()` I was triggering the UI thread checker. The only solution I can think of to work around this is to instantiate a fallback UIWindow on init and change this to:
```
return window ?? fallbackWindow
```

But then we would also need to enforce that `init` is always called on the main thread. I considered annotating the class with `@MainActor` but then it complicates the callsites where it gets instantiated.

Reply:
Yeah the only risk I see here is that since window is weak it could potentially become nil after checking it. That is unlikely and I can't think of a scenario where that could happen. That got me thinking though, do we really need to make window weak? As long as we nil out window after the continuation is complete I think it's reasonable to hold a strong reference to it.

Even better if you created a class dedicated to being the presentationContextProvider and have that take in a non optional then you will always be guaranteed a window and you don't need to worry about nil-ing anything out because that will happen automatically. The below is all just code I typed here to get the point across you might have to tweak it a bit:

```
class AuthenticatedWebViewPresentationContextProvider: NSObject, ASWebAuthenticationPresentationContextProviding {
   let window: UIWindow
   init(window: UIWindow) {
    self.window = window
   }

    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        return window
   }
 }

   class AuthenticatedWebViewManager: NSObject {
    ...

     func present(with url: URL, in window: UIWindow?) async throws -> URL? {
        guard authSession == nil else {
            throw AuthenticatedWebViewError.alreadyPresenting
        }
        guard let window else {
            throw AuthenticatedWebViewError.noWindow
        }

        // Context provider holds onto window but is only scoped to this function.
        let contextProvider = .init(window: window)

        let returnUrl: URL? = try await withCheckedThrowingContinuation { continuation in
            let authSession = sessionFactory(url, StripeConnectConstants.authenticatedWebViewReturnUrlScheme) { returnUrl, error in

                if let authenticationSessionError = error as? ASWebAuthenticationSessionError,
                   authenticationSessionError.code == .canceledLogin {
                    // The user either selected "Cancel" in the initial modal
                    // prompting them to "Sign In" or they hit the "Cancel"
                    // button in presented browser view
                    continuation.resume(returning: nil)
                    return
                } else if let error {
                    continuation.resume(throwing: error)
                    return
                }

                continuation.resume(returning: returnUrl)
            }
            authSession.presentationContextProvider = contextProvider

            self.authSession = authSession

            guard authSession.canStart,
                  authSession.start() else {
                continuation.resume(throwing: AuthenticatedWebViewError.cannotStartSession)
                return
            }
        }
        return returnUrl
    }
}


```

Reply:
The other option here is that present probably shouldn't take in a nil window if it's just going to immediately error. This seems like something that should be handled at the layer that is trying to present.

Reply:
yeah, I did that for conv

> The other option here is that present probably shouldn't take in a nil window if it's just going to immediately error. This seems like something that should be handled at the layer that is trying to present.

Yeah, I did that for convenience just so that I could throw one error and check for it in the ComponentWebViewController but probably not the best setup.

Reply:
> if you created a class dedicated to being the presentationContextProvider and have that take in a non optional then you will always be guaranteed a window

I like this idea. I'm gonna see if I can get this working

Reply:
- I took your suggestion of using a `AuthenticatedWebViewPresentationContextProvider` – much cleaner!
- I updated `present` to take a UIView and then grab its window param. I think this makes the `AuthenticatedWebViewError.noWindow` error handling make a lot more sense since the intent was to consolidate all the error handling into one place
- I annotated `present` with `@MainActor` to ensure we're always accessing the window on the main thread.